src/core/query.ts

```diff
@@ -450,6 +447,12 @@ export class Query<
       fn: context.fetchFn as () => TData,
       abort: abortController?.abort?.bind(abortController),
       onSuccess: data => {
+        if (typeof data === 'undefined') {
+          const error = new Error('Query data cannot be undefined') as any
+          this.onError(error)
```

Comment:
does this work?

```suggestion
          this.onError(new Error('Query data cannot be undefined') as TError)
```

Reply:
I tried that before. TS says:
> Conversion of type 'Error' to type 'TError' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  'TError' could be instantiated with an arbitrary type which could be unrelated to 'Error'.

So I need to cast to `any`. This makes sense, since errors coming in via try/catch are of type `any` (or `unknown` in newer TS versions). I wish JS had a real base Error type. Oh well.

Reply:
yep, TS is right here. The user can specify anything as `TError`, for example:

```
const { error } = useQuery<number, string>('key', () => Promise.resolve(5))

error?.toUppercase()
```

in that example, they would expect `error` to be a `string` - which it may or may not be. They would be in for a surprise if they then get an `Error`. We do the same thing already if `select` throws an error - we transform to `Error` state.

There are bunch of things we could do:
- ensure that `TError extends Error`. But you can reject a promise with non-Errors, too.
- remove the possibility to type `TError` completely and just always use `unknown`, which is the only "safe" thing to do. That would be pretty breaking for TS users though.

Jus thinking out loud. Nothing we need to do on this PR, the `any` cast is fine :)